File: forge/src/main/java/mcpc/tedo0627/kzeaddon/service/HidePlayerService.java
package mcpc.tedo0627.kzeaddon.service;

import net.minecraft.client.KeyMapping;
import net.minecraft.client.Minecraft;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.scores.Team;
import net.minecraftforge.client.event.InputEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;

import java.util.UUID;

public class HidePlayerService {

    private static boolean executing = false;

    public static boolean isInvisible(UUID uuid, Team team2, boolean bool) {
        if (team2 == null) return bool;

        Minecraft mc = Minecraft.getInstance();
        Player player = mc.player;
        if (player == null) return bool;
        if (player.getUUID().equals(uuid)) return bool;

        Team team1 = player.getTeam();
        if (team1 == null) return bool;

        if (!team1.getName().equals(team2.getName())) return bool;
        return executing || bool;
    }

    private final KeyMapping key;

    public HidePlayerService(KeyMapping key) {
        this.key = key;
    }

    @SubscribeEvent
    public void onKeyInput(InputEvent.KeyInputEvent event) {
        if (event.getKey() != key.getKey().getValue()) return;
        if (event.getAction() == 1 && key.isDown()) executing = !executing;
    }

    @SubscribeEvent
    public void onMouseInput(InputEvent.MouseInputEvent event) {
        if (event.getButton() != key.getKey().getValue()) return;
        if (event.getAction() == 1 && key.isDown()) executing = !executing;
    }
}

File: fabric/src/main/kotlin/mcpc/tedo0627/kzeaddon/fabric/service/HidePlayerService.kt
package mcpc.tedo0627.kzeaddon.fabric.service

import mcpc.tedo0627.kzeaddon.fabric.option.AddonOptions
import mcpc.tedo0627.kzeaddon.fabric.option.HideToggleType
import mcpc.tedo0627.kzeaddon.fabric.option.InvisibleType
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper
import net.minecraft.client.KeyMapping
import net.minecraft.client.Minecraft
import net.minecraft.world.scores.Team
import java.util.*

class HidePlayerService {

    companion object {

        private var executing = false

        /**
         * クリックして 半透明 -> 透明 -> 見える を切り替える時のフラグ
         * 半透明の時 false, 透明の時 true
         * */
        private var toggleInvisible = false

        @JvmStatic
        fun isInvisible(uuid: UUID, team: Team?, invisibleFlag: Boolean): Boolean {
            if (team == null) return invisibleFlag

            val mc = Minecraft.getInstance()
            val player = mc.player ?: return invisibleFlag
            if (player.uuid == uuid) return invisibleFlag
            if (player.team != team) return invisibleFlag

            return executing || invisibleFlag
        }

        @JvmStatic
        fun isOverrideIsInvisibleToFunc(): Boolean {
            if (!executing) return false

            if (AddonOptions.invisibleType.get() == InvisibleType.TOGGLE) {
                if (AddonOptions.hidePlayerToggle.get() != HideToggleType.SWITCH) return false

                return toggleInvisible
            }

            return AddonOptions.invisibleType.get() == InvisibleType.INVISIBLE
        }
    }

    private val key = KeyMapping("kzeaddon.key.hidePlayer", -1, "KZEAddon")

    init {
        KeyBindingHelper.registerKeyBinding(key)

        ClientTickEvents.END_CLIENT_TICK.register {
            when (AddonOptions.hidePlayerToggle.get()) {
                HideToggleType.SWITCH -> {
                    while (key.consumeClick()) {
                        if (AddonOptions.invisibleType.get() == InvisibleType.TOGGLE) {
                            if (executing && !toggleInvisible) { // 半透明の時
                                toggleInvisible = true
                            } else if (executing && toggleInvisible) { // 透明の時
                                executing = false
                                toggleInvisible = false
                            } else { // 見える時
                                executing = true
                            }
                        } else {
                            executing = !executing
                        }
                    }
                }
                HideToggleType.LONG_PRESS -> executing = key.isDown
                else -> throw IllegalStateException()
            }
        }
    }
}